### DISCUSSION TITLE
mouse via bluetooth

### DISCUSSION THREAD

**Discussion by beronk** [vAmigaWeb/vAmigaWeb#134] [Ideas]
Hi, is it possible to implement a mouse by bluetooth like the controllers?
To play some games by mouse instead by touchscreen?

**Comment by mithrendal** [comment]
Hi @beronk 

for real mouse input vAmigaWeb uses the standard pointerLock API when the input setting "mouse" is selected. This works for real mouse devices connected via cable and also bluetooth ... 

as far as I know the pointerLock API works fine only on non-mobile OS like macOS,windows,linux ... on mobile OS like android or iOS/iPadOS the pointerlockAPI is not implemented yet by apple or google...  even though you can physically connect/pair a mouse to it 

hopefully they will implement it someday ... 

wait let me check ... [caniuse.com](https://caniuse.com/pointerlock)

according to https://caniuse.com/pointerlock  mobile android meanwhile may work (I can not confirm because I have no test hardware for android) ... iOS still not supported 🙄 ... maybe iOS17 ?

**Comment by mithrendal** [comment]
@beronk 

When you put the following code snipped into a vAmigaWeb `action button` it will activate bluetooth mouse support 


```javascript

window.last_mouse_x=0;
window.last_mouse_y=0;
mouse_port=1;
has_pointer_lock_fallback=false;

function mouseDown(e) {
  Module._wasm_mouse_button(mouse_port,e.which, 1/* down */);
}
function mouseUp(e) {
  Module._wasm_mouse_button(mouse_port,e.which, 0/* up */);
}

function updatePosition_fallback(e) {
    let movementX=e.screenX-window.last_mouse_x;
    let movementY=e.screenY-window.last_mouse_y;
    window.last_mouse_x=e.screenX;
    window.last_mouse_y=e.screenY;
    let border_speed=4;
    let border_pixel=2;

    if(e.screenX<=border_pixel)
      movementX=-border_speed;
    if(e.screenX>=window.innerWidth-border_pixel)
      movementX=border_speed;
    if(e.screenY<=border_pixel)
      movementY=-border_speed;
    if(e.screenY>=window.innerHeight-border_pixel)
      movementY=border_speed;
 
 //   console.error(`avail=${window.screen.availWidth}x${window.screen.availHeight} mouse=${e.screenX}x${e.screenY} inner=${window.innerWidth}x${window.innerHeight}`);
    
    Module._wasm_mouse(mouse_port,movementX,movementY);  
}

window.add_pointer_lock_fallback=()=>{
  document.addEventListener("mousemove", updatePosition_fallback, false); 
  document.addEventListener("mousedown", mouseDown, false);
  document.addEventListener("mouseup", mouseUp, false);
};

window.request_pointerlock = async function() {
  if(canvas.requestPointerLock === undefined)
  {
    if(!has_pointer_lock_fallback)
    {
      add_pointer_lock_fallback();      
      has_pointer_lock_fallback=true;
    }
    return;
  }

  if(!has_pointer_lock && try_to_lock_pointer <20)
  {
    try_to_lock_pointer++;
    try {
      await canvas.requestPointerLock();           
      try_to_lock_pointer=0;
    } catch (error) {
      await sleep(100);
      await request_pointerlock();                
    }
  }
};
```

Tested on an iPad on iPadOS 16.3 with an Apple Magic Mouse Gen1

For reference here is a live viewer of all information we can currently get from a mouse event https://w3c.github.io/uievents/tools/mouse-event-viewer.html

What we need for the Amiga is movementX and movementY… but because this belongs to pointerlock and Apple doesn‘t give us currently the pointerlock we have to calculate or own relative delta to the last mouse movement (as you see in the code above) … its a bit hacky but it works 😎

**Comment by mithrendal** [comment]
Just pushed the above fallback implementation 😎 to https://vamigaweb.github.io in case pointerlock is not available on your system e.g. mobile safari

**Comment by beronk** [comment]
thanks ;) going to try it.

mithrendal ***@***.***> escreveu no dia sábado, 11/03/2023
à(s) 21:34:
